File: leetcode/c++/medium/_2429_minimize_xor.cpp
// Leetcode Problem 2429

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    int minimizeXor(int num1, int num2) {
        int x = 0;
        int bits = 0;

        while (num2 != 0) {
            bits += num2 & 1;
            num2 >>= 1;
        }

        for (int i = 32; i > 0; i--) {
            if ((num1 & (1 << (i-1))) != 0 && bits > 0) {
                x |= 1 << (i-1);
                bits--;
            }
        }

        for (int i = 0; i < 32; i++) {
            if (bits == 0) {
                break;
            }

            if ((x & (1 << i)) == 0) {
                x |= 1 << i;
                bits--;
            }
        }

        return x;
    }
};


class TestMinimizeXOR : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestMinimizeXOR, SimpleTestCase1) {
    EXPECT_EQ(solution.minimizeXor(3, 5), 3);
}

TEST_F(TestMinimizeXOR, SimpleTestCase2) {
    EXPECT_EQ(solution.minimizeXor(1, 12), 3);
}

TEST_F(TestMinimizeXOR, SimpleTestCase3) {
    EXPECT_EQ(solution.minimizeXor(15, 1), 8);
}

TEST_F(TestMinimizeXOR, SimpleTestCase4) {
    EXPECT_EQ(solution.minimizeXor(15, 15), 15);
}

File: leetcode/rust/problems/src/medium/_2429_minimize_xor.rs
// Leetcode Problem 2429

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn minimize_xor(num1: i32, mut num2: i32) -> i32 {
        let mut x: i32 = 0;

        let mut bits = 0;
        while num2 != 0 {
            bits += num2 & 1;
            num2 >>= 1;
        }

        // For each set bit in num1 set a bit in x if we have bits left to set
        for i in (0..32).rev() {
            if num1 & (1 << i) != 0 && bits > 0 {
                x |= 1 << i;
                bits -= 1;
            }
        }

        // Set the remaining bits
        for i in 0..32 {
            if bits == 0 {
                break;
            }

            if x & (1 << i) == 0 {
                x |= 1 << i;
                bits -= 1;
            }
        }

        x
    }
}


#[test]
fn test_minimize_xor() {
    assert_eq!(Solution::minimize_xor(3, 5), 3);
    assert_eq!(Solution::minimize_xor(1, 12), 3);
    assert_eq!(Solution::minimize_xor(15, 1), 8);
    assert_eq!(Solution::minimize_xor(15, 15), 15);
}